README.md
# Carousel Images

Carousel of images with reality (parallax) effect

![img](demo.gif)

## How to use

CarouselImages widget support Network and Asset Images. You should put to list path or url:
```
final List<String> listImages = [
  'assets/1.jpg',
  'https://www.factroom.ru/wp-content/uploads/2019/04/5-sobiraetes-syuda-priekhat.jpg',
  'https://cdn.flixbus.de/2018-01/munich-header-d8_0.jpg',
  'assets/4.jpg',
  'assets/5.jpg',
  ];'
```
```
CarouselImages(
  scaleFactor: 0.6,
  listImages: listImages,
  height: 300.0,
  borderRadius: 30.0,
  cachedNetworkImage: true,
  verticalAlignment: Alignment.topCenter,
  onTap: (index){
    print('Tapped on page $index');
  },
)
```

Parameter **height** is required!

Use **cachedNetworkImage** to cache images from network.

Use **scaleFactor** to give scale size of nearby images.
![img](scale.png)

Use **verticalAlignment** to place nearby images in right position.
![img](aligm.png)



CHANGELOG.md
## [1.0.0] - 14.06.2020

* initial release.

## [1.0.1] - 14.06.2020

* add description.

## [1.0.2] - 25.06.2020

* add borderRadius, cachedNetworkImage, scaleFactor, verticalAlignment.

## [1.0.3] - 27.06.2020

* change description.

## [1.1.0] - 02.06.2021

* null-safety support

## [1.1.1] - 29.06.2021

* dartfmt

## [1.1.2] - 15.07.2021

* add viewportFraction

lib/carousel_images.dart
library carouselimages;

import 'package:cached_network_image/cached_network_image.dart';
import 'package:flutter/material.dart';
import 'package:transparent_image/transparent_image.dart';
import 'dart:math' as math;

class CarouselImages extends StatefulWidget {
  ///List with assets path or url. Required
  final List<String> listImages;

  ///OnTap function. Index = index of active page. Optional
  final Function(int index)? onTap;

  ///Height of whole carousel. Required
  final double height;

  ///Possibility to cached images from network. Optional
  final bool cachedNetworkImage;

  ///Height of nearby images. From 0.0 to 1.0. Optional
  final double scaleFactor;

  ///Border radius of image. Optional
  final double? borderRadius;

  ///Vertical alignment of nearby images. Optional
  final Alignment? verticalAlignment;

  ///ViewportFraction. From 0.5 to 1.0. Optional
  final double viewportFraction;

  const CarouselImages({
    Key? key,
    required this.listImages,
    required this.height,
    this.onTap,
    this.cachedNetworkImage: false,
    this.scaleFactor = 1.0,
    this.borderRadius,
    this.verticalAlignment,
    this.viewportFraction = 0.9,
  })  : assert(scaleFactor > 0.0),
        assert(scaleFactor <= 1.0),
        super(key: key);

  @override
  _CarouselImagesState createState() => _CarouselImagesState();
}

class _CarouselImagesState extends State<CarouselImages> {
  late PageController _pageController;
  double _currentPageValue = 0.0;

  @override
  void initState() {
    super.initState();
    _pageController = PageController(
        viewportFraction: widget.viewportFraction.clamp(0.5, 1.0));
    _pageController.addListener(() {
      setState(() {
        _currentPageValue = _pageController.page!;
      });
    });
  }

  @override
  void dispose() {
    super.dispose();
    _pageController.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final width = MediaQuery.of(context).size.width;
    return SizedBox(
      height: widget.height,
      child: Material(
        color: Colors.transparent,
        child: AnimatedBuilder(
          animation: _pageController,
          builder: (context, child) {
            return PageView.builder(
              physics: BouncingScrollPhysics(),
              controller: _pageController,
              itemCount: widget.listImages.length,
              itemBuilder: (context, position) {
                double value = (1 -
                        ((_currentPageValue - position).abs() *
                            (1 - widget.scaleFactor)))
                    .clamp(0.0, 1.0);
                return Padding(
                  padding: EdgeInsets.symmetric(horizontal: 5.0),
                  child: Stack(
                    children: <Widget>[
                      SizedBox(
                          height: Curves.ease.transform(value) * widget.height,
                          child: child),
                      Align(
                        alignment: widget.verticalAlignment != null
                            ? widget.verticalAlignment!
                            : Alignment.center,
                        child: SizedBox(
                          height: Curves.ease.transform(value) * widget.height,
                          child: ClipRRect(
                            borderRadius: BorderRadius.circular(
                                widget.borderRadius != null
                                    ? widget.borderRadius!
                                    : 16.0),
                            child: Transform.translate(
                                offset: Offset(
                                    (_currentPageValue - position) *
                                        width /
                                        4 *
                                        math.pow(widget.viewportFraction, 3),
                                    0),
                                child: widget.listImages[position]
                                        .startsWith('http')
                                    ? widget.cachedNetworkImage
                                        ? CachedNetworkImage(
                                            imageUrl:
                                                widget.listImages[position],
                                            imageBuilder: (context, image) =>
                                                GestureDetector(
                                              onTap: () => widget.onTap != null
                                                  ? widget.onTap!(position)
                                                  : () {},
                                              child: Image(
                                                  image: image,
                                                  fit: BoxFit.fitHeight),
                                            ),
                                          )
                                        : GestureDetector(
                                            onTap: () => widget.onTap != null
                                                ? widget.onTap!(position)
                                                : () {},
                                            child: FadeInImage.memoryNetwork(
                                              placeholder: kTransparentImage,
                                              image:
                                                  widget.listImages[position],
                                              fit: BoxFit.fitHeight,
                                            ),
                                          )
                                    : GestureDetector(
                                        onTap: () => widget.onTap != null
                                            ? widget.onTap!(position)
                                            : () {},
                                        child: Image.asset(
                                          widget.listImages[position],
                                          fit: BoxFit.fitHeight,
                                        ),
                                      )),
                          ),
                        ),
                      )
                    ],
                  ),
                );
              },
            );
          },
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:carousel_images/carousel_images.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: HomePage(),
    );
  }
}

class HomePage extends StatelessWidget {
  final List<String> listImages = [
    'assets/1.jpg',
    'https://www.factroom.ru/wp-content/uploads/2019/04/5-osobennostej-klimata-pitera-o-kotoryh-vy-dolzhny-znat-esli-sobiraetes-syuda-priekhat.jpg',
    'https://cdn.flixbus.de/2018-01/munich-header-d8_0.jpg',
    'assets/4.jpg',
    'assets/5.jpg',
  ];

  @override
  Widget build(BuildContext context) {
    return Material(
      color: Colors.white,
      child: Column(
        children: <Widget>[
          SizedBox(height: 100),
          CarouselImages(
            scaleFactor: 0.7,
            listImages: listImages,
            height: 300.0,
            borderRadius: 30.0,
            cachedNetworkImage: true,
            verticalAlignment: Alignment.bottomCenter,
            onTap: (index) {
              print('Tapped on page $index');
            },
          )
        ],
      ),
    );
  }
}
